### DISCUSSION TITLE
Reference using a multivariable definition

### DISCUSSION THREAD

**Discussion by rhumbertgz** [eclipse-langium/langium#829] [Q&A] [answered]
Hello,

Could be possible to reference an object in the following oversimplified situation?

Line 1 defines three variables (x, y, z) of type "tello"
Line 2 the action 'go' reference one of the variables defined in l1

Example
--------------------
```
l1  let x, y, z tello
l2 go x
```

Grammar
---------------------
```
grammar HelloWorld

entry Model:
   (expression+=Expression)*;

Let: 'let' name+=ID (',' name+=ID)* type=('tello' | 'mbot');
Go: 'go' to=[Let:ID];

...
```

In this situation based on my current grammar definition I get an error saying that 'x' cannot be found.

Could be possible to reference a variable that was defined using the current syntax?

Best regards,
Humberto

**Comment by msujew** [comment, marked as answer]
Hey @rhumbertgz,

Langium is by default using a very simple lexical scoping mechanism (i.e. you can reference everything that is declared locally or in a parent of the current AST node). Since the `Let` and `Go` nodes probably live in different hierarchies, they can't see each other directly through the lexical scoping. So to fix this, you'll need to do the following:

1. First off, multiple names for single a element are not supported by default, but I believe you want to declare three separate elements anyway. So you'll need to refactor your `Let` rule slightly:
```
Let: 'let' variables+=Variable (',' variables+=Variable)* type=('tello' | 'mbot');
Variable: name=ID;
```
2. Tell the scope computation that you want to change the way your local scope mechanism works. We've just merged our [scoping guide](https://langium.org/guides/scoping/qualified-name/), where you'll find how to reference nodes from a different hierarchy using a fully qualified name. In your case, creating a FQN is not necessary, but you'll still need to override `computeLocalScopes` to create a local scope that not only incorporates the current element hierarchy but other elements as well.

> Could be possible to reference a variable that was defined using the current syntax?

You'll need to modify your grammar slightly, but that shouldn't be too much of an issue. Most of the work will be in correctly computing the local scope for the `Go.to` reference.

**Reply by rhumbertgz** [reply]
Thank you for the explanation @msujew
